worker/src/main/scala/org/apache/celeborn/service/deploy/worker/FetchHandler.scala

```diff
@@ -94,15 +94,168 @@ class FetchHandler(val conf: CelebornConf, val transportConf: TransportConf)
       case r: ChunkFetchRequest =>
         handleChunkFetchRequest(client, r)
       case r: RpcRequest =>
-        val msg = Message.decode(r.body().nioByteBuffer())
-        handleOpenStream(client, r, msg)
+        // 1. parse transportMessage
+        try {
+          val pbMsg = TransportMessage.fromByteBuffer(r.body().nioByteBuffer())
+          val pbOpenStream = pbMsg.getParsedPayload[PbOpenStream]
+          handleOpenStream(client, r.requestId, pbOpenStream, () => r.body().release())
+        } catch {
+          case _: Exception =>
+            // 2. parse old rpc message to keep client compatible
+            logDebug("use legacy format")
+            val msg = Message.decode(r.body().nioByteBuffer())
+            handleLegacyOpenStream(client, r, msg)
+        }
       case unknown: RequestMessage =>
         throw new IllegalArgumentException(s"Unknown message type id: ${unknown.`type`.id}")
     }
   }
 
-  // here are BackLogAnnouncement,OpenStream and OpenStreamWithCredit RPCs to handle
-  def handleOpenStream(client: TransportClient, request: RpcRequest, msg: Message): Unit = {
+  private def handleOpenStreamInternal(
+      client: TransportClient,
+      shuffleKey: String,
+      fileName: String,
+      startIndex: Int,
+      endIndex: Int,
+      initialCredit: Int,
+      creditStreamHandler: Consumer[java.lang.Long]): (Long, Int) = {
+    workerSource.startTimer(WorkerSource.OPEN_STREAM_TIME, shuffleKey)
+    var fileInfo = getRawFileInfo(shuffleKey, fileName)
+    fileInfo.getPartitionType match {
+      case PartitionType.REDUCE =>
+        if (endIndex != Integer.MAX_VALUE) {
+          fileInfo = partitionsSorter.getSortedFileInfo(
+            shuffleKey,
+            fileName,
+            fileInfo,
+            startIndex,
+            endIndex)
+        }
+        logDebug(s"Received chunk fetch request $shuffleKey $fileName " +
+          s"$startIndex $endIndex get file info $fileInfo")
+        if (fileInfo.isHdfs) {
+          (0, 0)
+        } else {
+          val buffers = new FileManagedBuffers(fileInfo, transportConf)
+          val fetchTimeMetrics = storageManager.getFetchTimeMetric(fileInfo.getFile)
+          val streamId = chunkStreamManager.registerStream(
+            shuffleKey,
+            buffers,
+            fetchTimeMetrics)
+          (streamId, fileInfo.numChunks())
+        }
+      case PartitionType.MAP =>
+        creditStreamManager.registerStream(
+          creditStreamHandler,
+          client.getChannel,
+          initialCredit,
+          startIndex,
+          endIndex,
+          fileInfo)
+        (-1, -1)
+      case PartitionType.MAPGROUP =>
+        (-1, -1)
+    }
+  }
+
+  def handleOpenStream(
+      client: TransportClient,
+      requestId: Long,
+      msg: PbOpenStream,
+      releaseFunc: () => Unit): Unit = {
```

Comment:
IMO using anonymous function is not a good style, just pass the `RpcRequest` and call r.body.release.